sdk/lib/cmlib/cmcheck.c

```diff
@@ -0,0 +1,1800 @@
+     * state structure. We'll be using it to walk
+     * down the registry hive tree in a recursive
+     * way without worrying that we explode the
+     * kernel stack in the most gruesome and gross
+     * ways.
+     */
+    WorkStateLength = CMP_REGISTRY_MAX_LEVELS_TREE_DEPTH * sizeof(CMP_REGISTRY_STACK_WORK_STATE);
+    WorkState = CmpAllocate(WorkStateLength,
+                            TRUE,
+                            TAG_REGISTRY_STACK);
+    if (!WorkState)
+    {
+        /* We're in despair if we can't get a buffer so bail out */
+        DPRINT1("CmpValidateRegistryInternal(): Couldn't allocate memory for registry stack work state!\n");
+        return CM_CHECK_REGISTRY_ALLOCATE_MEM_STACK_FAIL;
+    }
+
+    /* Obtain the root cell of the hive */
+    RootCell = GET_HHIVE_ROOT_CELL(Hive);
+    if (RootCell == HCELL_NIL)
+    {
+        /* Not getting the root cell is fatal, bail out */
+        DPRINT1("CmpValidateRegistryInternal(): Couldn't get the root cell of the hive!\n");
+        CmpFree(WorkState, WorkStateLength);
+        return CM_CHECK_REGISTRY_ROOT_CELL_NOT_FOUND;
+    }
+
+RestartValidation:
+    /*
+     * Prepare the stack state and start from
+     * the root cell. Ensure that the root cell
+     * itself is OK before we go forward.
+     */
+    StackDepth = 0;
+    WorkState[StackDepth].ChildCellIndex = 0;
+    WorkState[StackDepth].Current = RootCell;
+    WorkState[StackDepth].Parent = HCELL_NIL;
+    WorkState[StackDepth].Sibling = HCELL_NIL;
+
+    /*
+     * As we start checking the root cell which
+     * is the top element of a registry hive,
+     * we'll be going to look for child keys
+     * in the course of walking down the tree.
+     */
+    AllChildrenChecked = FALSE;
+
+    while (StackDepth >= 0)
+    {
+        /* Cache the current and parent cells */
+        CurrentCell = WorkState[StackDepth].Current;
+        ParentCell = WorkState[StackDepth].Parent;
+
+        /* Do we have still have children to validate? */
+        if (!AllChildrenChecked)
+        {
+            /* Check that the key is OK */
+            CmStatusCode = CmpValidateKey(Hive,
+                                          SecurityDefaulted,
+                                          ParentCell,
+                                          CurrentCell,
+                                          Flags);
+            if (!CM_CHECK_REGISTRY_SUCCESS(CmStatusCode))
+            {
+                /*
+                 * The key cell is damaged. We have to pray and
+                 * hope that this is not the root cell as any
+                 * damage done to the root is catastrophically
+                 * fatal.
+                 */
+                if (CurrentCell == RootCell)
+                {
+                    /*
+                     * IT IS THE ROOT!!! There's no hope for
+                     * this registry hive, it will go straight
+                     * to hell.
+                     */
```

Comment:
Nobody would miss that entire 5lines comment block. It doesn't transport any meaning that wouldn't be obvious already by the surrounding debug-prints and all the surrounding code.